File: packaging/mcpb/kotlin/stackql-mcp/src/main/kotlin/io/stackql/mcp/Platform.kt
package io.stackql.mcp

/**
 * Packaging platform keys. These match the stackql-mcpb-packaging release
 * asset names and the shared binary cache layout used by the npm and pypi
 * wrappers, so every StackQL distribution on a machine agrees on where a
 * given platform's binary lives.
 */
enum class Platform(val key: String) {
    LinuxX64("linux-x64"),
    LinuxArm64("linux-arm64"),
    WindowsX64("windows-x64"),
    DarwinUniversal("darwin-universal");

    /** The bundle asset name on the stackql release for this platform. */
    val bundleName: String get() = "stackql-mcp-$key.mcpb"

    /** Name the extracted server binary takes on this platform. */
    val exeName: String get() = if (this == WindowsX64) "stackql.exe" else "stackql"

    companion object {
        /**
         * Resolve the platform for the running JVM from `os.name`/`os.arch`.
         * Both macOS architectures map to the universal binary.
         */
        fun current(): Platform =
            detect(System.getProperty("os.name"), System.getProperty("os.arch"))

        /** Pure mapping, broken out so it is unit-testable without the JVM env. */
        fun detect(osNameRaw: String, osArch: String): Platform {
            val osName = osNameRaw.lowercase()
            val arch = normalizeArch(osArch.lowercase())
            return when {
                osName.contains("linux") && arch == "x64" -> LinuxX64
                osName.contains("linux") && arch == "arm64" -> LinuxArm64
                (osName.contains("windows")) && arch == "x64" -> WindowsX64
                (osName.contains("mac") || osName.contains("darwin")) &&
                    (arch == "x64" || arch == "arm64") -> DarwinUniversal
                else -> throw UnsupportedPlatformException(osName, osArch)
            }
        }

        private fun normalizeArch(osArch: String): String = when (osArch) {
            "amd64", "x86_64", "x64" -> "x64"
            "aarch64", "arm64" -> "arm64"
            else -> osArch
        }
    }
}

/** Thrown when no published stackql binary exists for the running platform. */
class UnsupportedPlatformException(osName: String, osArch: String) :
    StackqlMcpException("no published stackql binary for $osName/$osArch")

File: packaging/mcpb/go/embed/platform.go
package embed

import (
	"fmt"
	"runtime"
)

// Platform keys match the stackql-mcpb-packaging release asset names and the
// shared binary cache layout used by the npm and pypi wrappers.
const (
	PlatformLinuxX64        = "linux-x64"
	PlatformLinuxARM64      = "linux-arm64"
	PlatformWindowsX64      = "windows-x64"
	PlatformDarwinUniversal = "darwin-universal"
)

// PlatformKey returns the packaging platform key for the given GOOS/GOARCH
// pair. Both darwin architectures map to the universal binary.
func PlatformKey(goos, goarch string) (string, error) {
	switch goos {
	case "linux":
		switch goarch {
		case "amd64":
			return PlatformLinuxX64, nil
		case "arm64":
			return PlatformLinuxARM64, nil
		}
	case "windows":
		if goarch == "amd64" {
			return PlatformWindowsX64, nil
		}
	case "darwin":
		if goarch == "amd64" || goarch == "arm64" {
			return PlatformDarwinUniversal, nil
		}
	}
	return "", fmt.Errorf("stackql mcp: no published binary for %s/%s", goos, goarch)
}

// CurrentPlatformKey returns the platform key for the running program.
func CurrentPlatformKey() (string, error) {
	return PlatformKey(runtime.GOOS, runtime.GOARCH)
}

// exeName returns the name the extracted server binary uses inside the
// shared cache for the given platform key.
func exeName(platformKey string) string {
	if platformKey == PlatformWindowsX64 {
		return "stackql.exe"
	}
	return "stackql"
}
